src/builder.ts

```diff
@@ -83,10 +83,21 @@ export default class Builder {
 		// Connect the input stream to the rw stream
 		dup.setWritable(inputStream);
 
-		Bluebird.resolve(this.docker.buildImage(inputStream, buildOpts))
-			.then((res: NodeJS.ReadWriteStream) => {
-				let errored = false;
-				const outputStream = res
+		let streamError: Error;
+		const failBuild = _.once((err: Error) => {
+			streamError = err;
+			dup.destroy(err);
+			return this.callHook(hooks, 'buildFailure', handler, err, this.layers);
+		});
+
+		inputStream.on('error', failBuild);
+		dup.on('error', failBuild);
+
+		const buildPromise = Bluebird.try(() =>
+			this.docker.buildImage(inputStream, buildOpts),
+		).then((daemonStream: NodeJS.ReadStream) => {
+			return new Bluebird((resolve, reject) => {
```

Comment:
* `res` was renamed to `daemonStream`. It is the stream returned by dockerode's `buildImage`, connected to the output of the docker daemon.
* This `new Bluebird` promise is used to wait for the `outputStream.on('end')` event. Previously, the build was declared successful when the `end` event was received, regardless of the outcome of the promise returned by `buildStream` hook. See other comments below.